{
  "_url": "https://github.com/PowerShell/PowerShell/issues/17123",
  "author": "dolauli",
  "body": "### Prerequisites\n\n- [X] Write a descriptive title.\n- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)\n- [X] Search the existing issues.\n- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).\n- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).\n\n### Steps to reproduce\n\n1. Create a test.ps1 with following content.\r\n```\r\nfunction Test-error1 {\r\n    [CmdletBinding()]\r\n    param ()\r\n    process {\r\n        try {\r\n            Test-error2 @PSBoundParameters\r\n        } catch {\r\n            throw\r\n        }\r\n    }\r\n}\r\nfunction Test-error2 {\r\n    [CmdletBinding()]\r\n    param ()\r\n    process {\r\n        try {\r\n            write-error \"This is error#2\"\r\n        } catch {\r\n            throw\r\n        }\r\n    }\r\n}\r\n```\r\n2. run following command in PowerShell\r\n```\r\nPS > Import-Module .\\test.ps1\r\nPS > $ErrorActionPreference = 'Stop'\r\nPS > Test-error1 -ErrorVariable erroralias\r\nPS > $error\r\nPS > $erroralias\r\n```\n\n### Expected behavior\n\n```console\nThe output of $erroralias should be the same as $error as below.\r\n\r\nPS > $error\r\n\r\nTest-error2: C:\\Users\\dolauli\\source\\test\\test.ps1:33:13\r\nLine |\r\n  33 |              Test-error2 @PSBoundParameters\r\n     |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n     | This is error#2\n```\n\n\n### Actual behavior\n\n```console\nThe actually result is there are duplicated errors as below.\r\n\r\nPS > $erroralias\r\n\r\nErrorRecord                 : This is error#2\r\nWasThrownFromThrowStatement : True\r\nTargetSite                  : System.Collections.ObjectModel.Collection`1[System.Management.Automation.PSObject] Invoke(System.Collections.IEnumerable)\r\nMessage                     : The running command stopped because the preference variable \"ErrorActionPreference\" or common parameter is set to Stop: This is error#2\r\nData                        : {System.Management.Automation.Interpreter.InterpretedFrameInfo}\r\nInnerException              : \r\nHelpLink                    : \r\nSource                      : System.Management.Automation\r\nHResult                     : -2146233087\r\nStackTrace                  :    at System.Management.Automation.Runspaces.PipelineBase.Invoke(IEnumerable input)\r\n                                 at System.Management.Automation.Runspaces.Pipeline.Invoke()\r\n                                 at Microsoft.PowerShell.Executor.ExecuteCommandHelper(Pipeline tempPipeline, Exception& exceptionThrown, ExecutionOptions options)\r\n\r\n\r\nTest-error2: C:\\Users\\dolauli\\source\\test\\test.ps1:33:13\r\nLine |\r\n  33 |              Test-error2 @PSBoundParameters\r\n     |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n     | This is error#2\r\n\r\n\r\nTest-error2: C:\\Users\\dolauli\\source\\test\\test.ps1:33:13\r\nLine |\r\n  33 |              Test-error2 @PSBoundParameters\r\n     |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n     | This is error#2\r\n\r\n\r\nTest-error2: C:\\Users\\dolauli\\source\\test\\test.ps1:33:13\r\nLine |\r\n  33 |              Test-error2 @PSBoundParameters\r\n     |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n     | This is error#2\r\n\r\n\r\nTest-error2: C:\\Users\\dolauli\\source\\test\\test.ps1:33:13\r\nLine |\r\n  33 |              Test-error2 @PSBoundParameters\r\n     |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n     | This is error#2\r\n\r\n\r\nTest-error2: C:\\Users\\dolauli\\source\\test\\test.ps1:33:13\r\nLine |\r\n  33 |              Test-error2 @PSBoundParameters\r\n     |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n     | This is error#2\r\n\r\n\r\nTest-error2: C:\\Users\\dolauli\\source\\test\\test.ps1:33:13\r\nLine |\r\n  33 |              Test-error2 @PSBoundParameters\r\n     |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n     | This is error#2\r\n\r\n\r\nPS C:\\Users\\dolauli\\source\\test>\n```\n\n\n### Error details\n\n_No response_\n\n### Environment data\n\n```powershell\nName                           Value\r\n----                           -----\r\nPSVersion                      7.2.2\r\nPSEdition                      Core\r\nGitCommitId                    7.2.2\r\nOS                             Microsoft Windows 10.0.19043\r\nPlatform                       Win32NT\r\nPSCompatibleVersions           {1.0, 2.0, 3.0, 4.0\u2026}       \r\nPSRemotingProtocolVersion      2.3\r\nSerializationVersion           1.1.0.1\r\nWSManStackVersion              3.0\n```\n\n\n### Visuals\n\nWhen $ErrorActionPreference = 'Continue', it works as expected.\r\n![image](https://user-images.githubusercontent.com/11451757/162346033-4acefba7-5134-4910-9206-705be9b1388a.png)\r\n\r\nWhen $ErrorActionPreference = 'Stop', there will be duplicated errors.\r\n![image](https://user-images.githubusercontent.com/11451757/162346203-fbaf5ef0-1e78-4d7f-a534-51179125a044.png)\r\n\r\n",
  "closed_at": null,
  "comments": [
    {
      "author": "jhoneill",
      "author_association": "NONE",
      "body": "Here is a simpler repro\r\n```\r\n#501 PS7 ~\\Documents\\PowerShell> function foo {1/0}\r\n#502 PS7 ~\\Documents\\PowerShell> $Error.Count\r\n0\r\n#503 PS7 ~\\Documents\\PowerShell> foo\r\nRuntimeException: Attempted to divide by zero.\r\n#504 PS7 ~\\Documents\\PowerShell> $ErrorActionPreference=\"Stop\"\r\n#505 PS7 ~\\Documents\\PowerShell> foo\r\nParentContainsErrorRecordException: Attempted to divide by zero.\r\n#506 PS7 ~\\Documents\\PowerShell> $Error.Count\r\n3\r\n#507 PS7 ~\\Documents\\PowerShell> $Error[0].GetType()\r\n\r\nIsPublic IsSerial Name                                     BaseType\r\n-------- -------- ----                                     --------\r\nTrue     True     RuntimeException                         System.SystemException\r\n\r\n#508 PS7 ~\\Documents\\PowerShell> $Error[1].GetType()\r\n\r\nIsPublic IsSerial Name                                     BaseType\r\n-------- -------- ----                                     --------\r\nTrue     True     ErrorRecord                              System.Object\r\n\r\n#509 PS7 ~\\Documents\\PowerShell> $Error[2].GetType()\r\n\r\nIsPublic IsSerial Name                                     BaseType\r\n-------- -------- ----                                     --------\r\nTrue     True     ErrorRecord                              System.Object\r\n```\r\n\r\nThe error itself adds an _error record_ to `$error` , and to anything specified in `-errorVariable` \r\n\r\nThe ErrorActionPreference means the error generates a _runtime exception_ which is logged separately to $error and will be the more recent event there.  \r\n\r\n **However** specifying ErrorAction as a switch for the function does not log a duplicate - see below   \r\n preference as continue, no commandline switch adds 1 error. \r\n preference as continue, with commandline switch adds 1 error. \r\n preference as stop, no commandline switch adds 2 errors. \r\n preference as stop, with commandline switch adds 2 errors.  \r\n\r\n```\r\n#501 PS7 ~\\Documents\\PowerShell> function foo {[cmdletbinding()] param() 1/0 ; 1/2}\r\n#502 PS7 ~\\Documents\\PowerShell> foo\r\nRuntimeException: Attempted to divide by zero.\r\n0.5\r\n#503 PS7 ~\\Documents\\PowerShell> $error.count\r\n1\r\n#504 PS7 ~\\Documents\\PowerShell> foo -ErrorAction stop\r\nfoo: Attempted to divide by zero.\r\n#505 PS7 ~\\Documents\\PowerShell> $error.count\r\n2\r\n#506 PS7 ~\\Documents\\PowerShell> $ErrorActionPreference=\"Stop\"\r\n#507 PS7 ~\\Documents\\PowerShell> foo\r\nfoo: Attempted to divide by zero.\r\n#508 PS7 ~\\Documents\\PowerShell> $error.count\r\n4\r\n#509 PS7 ~\\Documents\\PowerShell> foo -ErrorAction Stop\r\nfoo: Attempted to divide by zero.\r\n#510 PS7 ~\\Documents\\PowerShell> $error.count\r\n6\r\n```\r\n\r\n**Either** the erroractionPreference SHOULD NOT ADD the second entry,  \r\n**Or** the -ErrorAction switch  SHOULD ADD it. \r\n\r\nBut it isn't consistent.  It probably can be made consistent with low risk of breaking things, but needs more thought and discussion.\r\n\r\n\r\n",
      "created_at": "2022-04-08T09:12:56Z",
      "updated_at": "2022-04-08T09:15:36Z"
    }
  ],
  "created_at": "2022-04-08T01:39:29Z",
  "labels": [
    "WG-Engine",
    "Needs-Triage"
  ],
  "number": 17123,
  "state": "open",
  "title": "Duplicated errors when using $ErrorEnvirable with $ErrorActionPreference = 'Stop'",
  "updated_at": "2022-04-08T13:52:04Z"
}